tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -2400,6 +2453,97 @@ static CXXRecordDecl *AddBuiltInTriangleIntersectionAttributes(ASTContext& conte
+// Subobjects
+
+static CXXRecordDecl *StartSubobjectDecl(ASTContext& context, const char *name) {
+  IdentifierInfo &id = context.Idents.get(StringRef(name), tok::TokenKind::identifier);
+  CXXRecordDecl *decl = CXXRecordDecl::Create( context, TagTypeKind::TTK_Struct, 
+    context.getTranslationUnitDecl(), NoLoc, NoLoc, &id, nullptr, DelayTypeCreationTrue);
+  decl->startDefinition();
+  return decl;
+}
+
+void FinishSubobjectDecl(ASTContext& context, CXXRecordDecl *decl) {
+  decl->completeDefinition();
+  context.getTranslationUnitDecl()->addDecl(decl);
+  decl->setImplicit(true);
+}
+
+// struct StateObjectConfig 
+// {
+//   uint32_t Flags;
+// };
+static CXXRecordDecl *CreateSubobjectStateObjectConfig(ASTContext& context) {
+  CXXRecordDecl *decl = StartSubobjectDecl(context, "StateObjectConfig");
+  CreateSimpleField(context, decl, "Flags", context.UnsignedIntTy, AccessSpecifier::AS_private);
+  FinishSubobjectDecl(context, decl);
+  return decl;
+}
+
+// struct GlobalRootSignature
+// {
+//   string signature;
+// };
+static CXXRecordDecl *CreateSubobjectRootSignature(ASTContext& context, bool global) {
+  CXXRecordDecl *decl = StartSubobjectDecl(context, global ? "GlobalRootSignature" : "LocalRootSignature");
+  CreateSimpleField(context, decl, "Data", context.HLSLStringTy, AccessSpecifier::AS_private);
+  FinishSubobjectDecl(context, decl);
+  return decl;
+}
+
+// struct SubobjectToExportsAssociation 
+// {
+//   string Subobject;
+//   string Exports;
+// };
+static CXXRecordDecl *CreateSubobjectSubobjectToExportsAssoc(ASTContext& context) {
+  CXXRecordDecl *decl = StartSubobjectDecl(context, "SubobjectToExportsAssociation");
+  CreateSimpleField(context, decl, "Subobject", context.HLSLStringTy, AccessSpecifier::AS_private);
+  CreateSimpleField(context, decl, "Exports",   context.HLSLStringTy, AccessSpecifier::AS_private);
+  FinishSubobjectDecl(context, decl);
+  return decl;
+}
+
+// struct RaytracingShaderConfig 
+// {
+//   uint32_t MaxPayloadSizeInBytes;
+//   uint32_t MaxAttributeSizeInBytes;
+// };
+static CXXRecordDecl *CreateSubobjectRaytracingShaderConfig(ASTContext& context) {
+  CXXRecordDecl *decl = StartSubobjectDecl(context, "RaytracingShaderConfig");
+  CreateSimpleField(context, decl, "MaxPayloadSizeInBytes",   context.UnsignedIntTy, AccessSpecifier::AS_private);
+  CreateSimpleField(context, decl, "MaxAttributeSizeInBytes", context.UnsignedIntTy, AccessSpecifier::AS_private);
+  FinishSubobjectDecl(context, decl);
+  return decl;
+}
+
+// struct RaytracingPipelineConfig 
+// {
+//   uint32_t MaxTraceRecursionDepth;
+// };
+static CXXRecordDecl *CreateSubobjectRaytracingPipelineConfig(ASTContext& context) {
+  CXXRecordDecl *decl = StartSubobjectDecl(context, "RaytracingPipelineConfig"); 
+  CreateSimpleField(context, decl, "MaxTraceRecursionDepth", context.UnsignedIntTy, AccessSpecifier::AS_private);
+  FinishSubobjectDecl(context, decl);
+  return decl;
+}
+
+// struct HitGroup
+// {
+//   string intersection;
```

Comment:
I've been considering reordering this to match the D3D12 runtime structure, with the intersection shader listed last.  There, it's anyhit, closesthit, then intersection.  We will want to reorder it everywhere though, and should do so right away before it starts getting used.